src/runtime/ops/ecs/value.rs

```diff
@@ -244,22 +362,53 @@ pub fn ecs_value_ref_set(
     // Access the provided path on the value ref
     let mut value_ref = append_path(value_ref, path, world)?;
 
-    // Get the reflect value
-    let mut reflect = value_ref.get_mut(world).unwrap();
-
-    // Try to store a primitive in the value
-    try_downcast_leaf_set!(reflect <- new_value for
-        u8, u16, u32, u64, u128, usize,
-        i8, i16, i32, i64, i128, isize,
-        String, char, bool, f32, f64
-    )
-    .map(|_| serde_json::Value::Null)
-    .map_err(|e| {
-        format_err!(
-            "could not set value reference: type `{typename}` is not a primitive type: {e}",
-            typename = reflect.type_name(),
+    // Try to asign as a primitive
+    let primitive_assignment_result = {
+        let new_value = new_value.clone();
+        let mut reflect = value_ref.get_mut(world)?;
+
+        // Try to store a primitive in the value
+        try_downcast_leaf_set!(reflect <- new_value for
+            u8, u16, u32, u64, u128, usize,
+            i8, i16, i32, i64, i128, isize,
+            String, char, bool, f32, f64
         )
-    })
+        .map_err(|e| {
+            format_err!(
+                "could not set value reference: type `{type_name}` is not a primitive \
+                type or value ref: {e}",
+                type_name = reflect.type_name(),
+            )
+        })
+    };
+
+    // If we could not assign a primitive
+    if let Err(e) = primitive_assignment_result {
+        // Try to assign as a reflect value
+        if let Ok(new_js_value_ref) = serde_json::from_value::<JsValueRef>(new_value) {
+            let new_value_ref = value_refs
+                .get(new_js_value_ref.key)
+                .ok_or_else(|| format_err!("Value ref doesn't exist"))?;
+            let new_reflect = new_value_ref.get(world)?.clone_value();
```

Comment:
That does print 5 just like you are expecting.

The clone is actually simply to avoid borrowing issues, because we aren't allowed to take an immutable reference to `world` when running `new_value_ref.get(world)` and then hold that ref while we mutably borrow world with `value_ref.get_mut(world)`.

It doesn't change semantics, but it does obviously result in an extra clone as far as performance is concerned.

I was wondering whether or not there would be a way to somehow safely borrow multiple value refs from the world at the same time, having it make sure the borrows don't alias when you do the `borrow_multiple()` or something like that.